### PR TITLE
Fix flaky e2e: Use expectation model for deployment's new rc creation

### PR DESCRIPTION
**Pull request by janetkuo**
Attempt to fix flaky deployment e2e tests #18996 #19064 #19299

In the failed cases, multiple rcs are created by a deployment and messes things up. Now we use expectation model for new rc creation. 

cc @bprashanth @nikhiljindal @ghodss @ironcladlou @kargakis @bgrant0607

### CODE REVIEW

File: pkg/controller/deployment/deployment_controller.go [hunk 1]
```diff
@@ -78,7 +78,11 @@ type DeploymentController struct {
 	podStoreSynced func() bool
 
 	// A TTLCache of pod creates/deletes each deployment expects to see
-	expectations controller.ControllerExpectationsInterface
+	podExpectations controller.ControllerExpectationsInterface
+
+	// A TTLCache of rc creates/deletes each deployment expects to see
+	// TODO: make expectation model understand (rc) updates (besides adds and deletes)
+	rcExpectations controller.ControllerExpectationsInterface
```
Comment (bprashanth) [inline pkg/controller/deployment/deployment_controller.go]: suggest also renaming the variables in controller_utils that indicate that expectations are only used for pods

File: pkg/controller/deployment/deployment_controller.go [hunk 2]
```diff
@@ -411,6 +423,16 @@ func (dc *DeploymentController) syncRecreateDeployment(deployment extensions.Dep
 }
 
 func (dc *DeploymentController) syncRollingUpdateDeployment(deployment extensions.Deployment) error {
+	// Check the rc expectations of deployment before getting new rc
+	dKey, err := controller.KeyFunc(&deployment)
+	if err != nil {
+		return fmt.Errorf("Couldn't get key for deployment %#v: %v", deployment, err)
+	}
+	if !dc.rcExpectations.SatisfiedExpectations(dKey) {
+		fmt.Printf("RC expectations not met yet before getting new RC\n")
```
Comment (bprashanth) [inline pkg/controller/deployment/deployment_controller.go]: use glog.V(4) instead?

**Comment by bprashanth** [comment]
Suggest adding a unittest. Also suggest following up on the todo when you have time, or just centralizing the handling of expectations. The original goal of using expectations was to avoid hand management of locks across the controller which could be very confusing, as we use multiple expectations caches we approach that model more and more.

**Comment by janetkuo** [comment]
> Suggest adding a unittest.

@bprashanth I'm not sure a unit test for which part?

**Comment by bprashanth** [comment]
To make sure the controller waits till rc and pod expectations are satisfied before doing anything (maybe? unsure if that's the best thing to test); essentially something should consistently break (so not e2e flake) if someone removes or wrongfully sets either expectation. But this is just a suggestion, if you don't think it's worth the effort I leave it to you.

File: pkg/controller/deployment/deployment_controller.go (hunk 1 above)
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: Done

**Comment by janetkuo** [comment]
Updated controller_utils.go to generalize the name of expectations variables. I also made a little change to do expectation check after trying to get existing new rc and before creating new rc (the PR originally does the check before getting existing new rc). 

@bprashanth I'd love to add unit tests for pod and rc expectations after finishing 1.2 Deployment features.

**Comment by bprashanth** [comment]
LGTM

**Comment by janetkuo** [comment]
Rebased.

**Pull request merged**